### DISCUSSION TITLE
How to use Testcontainers and Dependencie Injection ?

### DISCUSSION THREAD

**Discussion by Samyssmile** [quarkusio/quarkus#28154] [Q&A] [answered]
I try to migrate my Integration Test Class to use @Testcontainers.

Original Test Class was like

    @QuarkusTest
    class GameResourceTest {
    
        @Inject
        TeamService teamService;
    
        @Test
        void shouldLeadTheRankLadder() {
         teamService.doCrazyStuff();  // PASS

After rewrite it looks like this

    @Testcontainers
    class GameResourceTest {
    
        @Container
        private MariaDBContainer mariaDBContainer = new MariaDBContainer("mariadb:10.5.16").withDatabaseName("test").withUsername("test").withPassword("test");
        
        @Inject
        TeamService teamService;
    
        @Test
        void test() {
         assertTrue(mariaDBContainer.isRunning()); // PASS
        }
    
        @Test
        void shouldLeadTheRankLadder() {
         teamService <-----------------------IS NULL HERE

So after removing the @QuarkusTest annoation, dependency injection of my service not work anymore.

How to use Testcontainers and Dependency Injection here?

[a comment by an automated account is not shown]

**Comment by manovotn** [comment, marked as answer]
`@QuarkusTest` is a Quarkus JUnit extension which takes care of Quarkus startup and provides the test instance which is injected into.
If you remove this extension and replace it with `@Testcontainers` (which is provided by JUnit and not related to Quarkus in any way), then Quarkus functionality will disappear.

Also, I don't see any tests in Quarkus that would use `@Testcontainers` ATM so I'd say this isn't supported.
Am I right @geoand?

**Reply by geoand** [reply]
That is correct. 

I have answered the same question on [SO](https://stackoverflow.com/a/73810045/2504224) BTW

**Reply by mkouba** [reply]
I think that you'll need a custom `QuarkusTestLifecycleManager`, see https://quarkus.io/guides/getting-started-testing#launching-containers for more info. FYI I've found this video from Alex Soto Bueno about Quarkus and Testcontainers: https://www.youtube.com/watch?v=xtQ1GmVpPP0 . It might be outdated though - made in 2019 ;-).

**Reply by Samyssmile** [reply]
Ok, I solved it with QuarkusTestLifecycleManager. I missundertood the @Testcontainers Annotation. For me it was like 'add @Testcontainers and define MariaDB will force the test automaticly use the Testcontainer  for all transactions.

With QuarkusTestLifecycleManager it works as expected. This is closed. Thank you
